design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,125 @@
+
+Lambda functions will also enable some users to use ksqlDB in environments that do not allow for the deployment of UDFs.
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
+
+```
+arg => expr
+```
+
+This expression represents the following:
+
+* `arg` is the argument passed into the scope of each invocation of the lambda function.
+* `expr` is the "body" of the lambda function. Each invocation of the lambda function will evaluate this expression and return the result.
+
+Multiple arguments should also be allowed for lambda functions. When multiple arguments are used, they should be wrapped in parentheses:
+
+```
+(x, y) => expr
+```
+
+#### Allowed expressions
+
+The body of each lambda function should simply be **any expression that could be used in a `SELECT` expression list**, with some exceptions:
+
+* Aggregate function calls (e.g. `sum`, `count`, `avg`) should not be allowed (please see *[Open questions](#open-questions)* section). Note that simple aggregations are still possible with the `reduce_*` invocation functions.
+* `SELECT` subqueries should not be allowed.
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_keys(map, k => k)` - Applies the given lambda function to each key in the input `Map`, returning a new `Map` containing the transformed keys.
+
+* `transform_values(map, v => v)` - Applies the given lambda function to each value in the input `Map`, returning a new `Map` containing the transformed values.
+
+* `transform_map(map, (k, v) => new_k, (k, v) => new_v)` - Applies the given lambda functions to the keys and values of the input `Map`, respectively. A new `Map` containing the transformed keys and values is returned.
```

Comment:
For functions that may result in ambiguous types of `new_k, new_v` e.g. such that KSQL cannot infer the new schema field types, would we require users to specify the return value types in their lambda function? If yes, how we would enforce this?